firmware/library/L1_Drivers/timer.hpp

```diff
@@ -0,0 +1,157 @@
+    kRestartStop          = 0b110,
+    kInterruptRestartStop = 0b111
+  };
+  virtual void Initialize(TimerPort timer)= 0;
+  virtual void SetTimer(uint32_t time, RegisterChoice reg,
+                        TimerSetting mode) = 0;
+  virtual uint32_t GetTimer() = 0;
+  virtual void ClearInterrupts() = 0;
+};
+
+class Timer : public TimerInterface, protected Lpc40xxSystemController
+{
+ public:
+  static constexpr uint8_t kTime0Mat0   = 0b011;
+  static constexpr uint8_t kClearMode   = 0b0111;
+  static constexpr uint8_t kPrescaleMicro   = 48;
+  static constexpr uint32_t kClear = 0b11111111111111111111111111111111;
+
+  enum TimerInterruptBits : uint8_t
+  {
+    kMat0 = 0,
+    kMat1 = 3,
+    kMat2 = 6,
+    kMat3 = 9
+  };
+
+  inline static Pin quads[4][4] = {
+                  { Pin(1, 28), Pin(1, 29), Pin(0, 15), Pin(0, 15) },
+                  { Pin(1, 22), Pin(1, 25), Pin(0, 3),  Pin(0, 3) },
+                  { Pin(2, 5) , Pin(2, 4),  Pin(2, 3),  Pin(2, 2) },
+                  { Pin(2, 26), Pin(2, 27), Pin(2, 30), Pin(2, 31) } };
+//                  MAT 0     , Mat 1     , Mat 2     , Mat 3
+
+  explicit constexpr Timer(TimerPort mode)
+                      : channel_(static_cast<uint8_t>(mode)),
+                        mat0_(&quads[channel_][0]),
+                        mat1_(&quads[channel_][1]),
+                        mat2_(&quads[channel_][2]),
+                        mat3_(&quads[channel_][3])
+  {
+  }
+
+  void Initialize(TimerPort timer) override
+  {
+    PowerUpPeripheral(Lpc40xxSystemController::Peripherals::kTimer0);
+    PowerUpPeripheral(Lpc40xxSystemController::Peripherals::kTimer1);
+    PowerUpPeripheral(Lpc40xxSystemController::Peripherals::kTimer2);
+    PowerUpPeripheral(Lpc40xxSystemController::Peripherals::kTimer3);
+    mat0_->SetPinFunction(kTime0Mat0);
+    mat0_->SetAsAnalogMode(true);
+    mat0_->SetMode(PinInterface::Mode::kInactive);
+    mat1_->SetPinFunction(kTime0Mat0);
+    mat1_->SetAsAnalogMode(true);
+    mat1_->SetMode(PinInterface::Mode::kInactive);
+    mat2_->SetPinFunction(kTime0Mat0);
+    mat2_->SetAsAnalogMode(true);
+    mat2_->SetMode(PinInterface::Mode::kInactive);
+    mat3_->SetPinFunction(kTime0Mat0);
+    mat3_->SetAsAnalogMode(true);
+    mat3_->SetMode(PinInterface::Mode::kInactive);
+    // set Prescale register for Prescale Counter to milliseconds
+    tim_register[static_cast<uint8_t>(timer)]->PR &= ~(kClear << 1);
+    tim_register[static_cast<uint8_t>(timer)]->PR |= (kPrescaleMicro << 1);
+    tim_register[static_cast<uint8_t>(timer)]->TCR |= (1 << 0);
+  }
+
+  // Function that sets a timer of your choice (0-3) to a specific
+  // time in milliseconds,and designates the mode of
+  // functionality you want from it, interrupt, stop, or reset.
+  // --NOTICE-- Currently Each Timer is directly hooked to match
+  void SetTimer(uint32_t time, RegisterChoice reg, TimerSetting mode) override
+  {
+        switch (reg)
+    {
+        case kRegMR0:
```

Comment:
Use a lookup table rather then copying the same lines of code 4 times. The are all the same except that you shift by kMatX.

Reply:
So i am trying to use a lookup table for the pointers to the match registers and I have tried to do it like this
`static volatile uint32_t * match[6] = 
               { &tim_register[channel_]->MR0, &tim_register[channel_]->MR1,
                 &tim_register[channel_]->MR2, &tim_register[channel_]->MR3,
                 &tim_register[channel_]->CR0, &tim_register[channel_]->CR1
               };`

However It says that channel_ is not declared in this scope.  I am sorry to ask you, but I am confused as to how I can create this lookup table or where I have made a mistake.